packages/ingest/recursion.js

```diff
@@ -54,28 +42,37 @@ async function recurOnDirectory(fn, currentPath, segments, position) {
  * It requests a promisified list function that returns contents of
  * a directory on a server, filtering on provided regex segments.
  *
- * Note that calls to the list function will not have leading or terminating slashes.
- * Initially an empty string is passed as the path to list the default directory. Following calls
- * based on items discovered will be of the format `fn('path/to/files')`, again with no leading or
- * terminating slashes.
+ * Note that calls to the list function will use either a relative or absolute path, corresponding
+ * to the `configuredPath` passed into this function. The list function will initially be called
+ * with '.' for a relative path or '/' for an absolute path. List functions will need to be able to
+ * normalize or correct these paths as appropriate for their protocol.
+ *
+ * Further calls to the list functions will append the current path to that starting path, such
+ * that all calls will start with either '.' or '/', regardless of additional characters, e.g.
+ * `fn('./path')` vs. `fn('path')`.
  *
- * List functions will need to be able to normalize or correct these paths as appropriate for their
- * protocol.
+ * In the case of failure during the recursive list, this function will only apply `path.normalize`
+ * to the `configuredPath` and then call the list function with the entire normalizedPath.
  *
  * @param {function} fn - the promisified function for listing a directory
- * @param {string} originalPath - path string which may contain regexes for filtering
+ * @param {string} configuredPath - path string configured by operator, which may contain
+ *                                  regexes for filtering
  * @returns {Promise} the promise of an object that has the path is the key and
  *   list of files as values
  */
-async function recursion(fn, originalPath) {
-  const normalizedPath = normalizeSlashes(originalPath);
+async function recursion(fn, configuredPath) {
+  const normalizedPath = path.normalize(configuredPath);
+  const isAbsolutePath = path.isAbsolute(normalizedPath);
   try {
-    const segments = normalizedPath.split('/'); // split on divider
-    return await recurOnDirectory(fn, segments[0], segments, 0);
+    const segments = normalizedPath
+      .split('/') // split on divider
+      .filter((segment) => segment.trim() !== ''); // filter out empty strings from split
+    const startingPath = isAbsolutePath ? '/' : '.';
```

Comment:
I'm not sure that I understand this behavior and/or how it differs from the previous behavior. If I pass in `/path/to/dir` as my `configuredPath`, then why do we start by trying to list files in `/`? In other words, why would we start in parent directories of the directory that I've specified?

Reply:
Because with the regex filtering, you could pass in `/path/to/dir` or you could pass in `/(p.*)/to/dir`. In the latter case you want to list everything in `/` and recur into everything that matches `(p.*)`.

In the previous implementation it kind of haphazardly tried to figure out which part is text and which part is regex and it was highly confusing and not maintainable, as well as broken. I realize there may be problems with permissions when trying to list parent directories so that's why the fallback _does_ start by listing your `configuredPath`.

Reply:
>  it was highly confusing and not maintainable, as well as broken

Well when you put it _that_ way...